### PR TITLE
Campaign message queue

### PR DESCRIPTION
**Pull request by mqueme**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/103
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR changes the way marketing emails are handled at the moment.  With this, when a campaign email is triggered the email will be sent to a queue to be processed instead of sending it right away.  This is due to the fact of a previous feature introduced in version 2.1.0. Frequency rules for contacts.  

Basically, the way this is handling email sending at the moment, the email is lost if the contact has met its current frequency rules. With this, a message can be rescheduled if a contact was not contacted due to frequency rules, also if a message was failed then, it tries to send again until the max attempts are met or the message was sent successfully.

#### Steps to test this PR:
1. Apply this PR
2. create a new campaign and add a new email to be sent as an action. Make sure you select marketing emails for the queue configuration appears.
3. fill in the priority and retry times. Save your campaign.
4. trigger your campaign and see a new message being inserted in the message_queue table.
5. create this new cron to run every so often and send your messages. like this 
6. This PR also bypasses the transactional email to be sent immediately without checking for contact frequency rules. 

php app/console mautic:campaigns:messagequeue

email messages should be sent or rescheduled, marked as sent, and other data.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by escopecz** [comment]
I got stuck at
> trigger your campaign and see a new message being inserted in the message_queue table.

It doesn't create any rows in the table. I'll test again from scratch to be sure it's not user error.

**Comment by mqueme** [comment]
@escopecz send email was not being called correctly after changing the insert into queue from position.

**Comment by dongilbert** [comment]
Retested, works well. Thanks @mqueme

File: app/migrations/Version20160916000000.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+     */
+    public function up(Schema $schema)
+    {
+        $leadIDX  = $this->generatePropertyName('message_queue', 'idx', ['lead_id']);
+        $leadFK   = $this->generatePropertyName('message_queue', 'fk', ['lead_id']);
+        $campaignFK    = $this->generatePropertyName('message_queue', 'idx', ['campaign_id']);
+
+        $sql = <<<SQL
+CREATE TABLE {$this->prefix}message_queue (
+  `id` int(10) unsigned NOT NULL,
```
Comment (escopecz) [inline app/migrations/Version20160916000000.php]: This must be `AUTO_INCREMENT`

**Comment by escopecz** [comment]
If I execute the 
```
PHP Notice:  Undefined index: sms in app/bundles/CoreBundle/Model/MessageQueueModel.php on line 122
and
PHP Warning:  DateTime::add() expects parameter 1 to be DateInterval, string given in app/bundles/CoreBundle/Model/MessageQueueModel.php on line 143
```

**Pull request merged**